src/WebJobs.Script/Host/FunctionMetadataManager.cs

```diff
@@ -18,36 +21,95 @@ namespace Microsoft.Azure.WebJobs.Script
+        public ImmutableArray<FunctionMetadata> GetFunctionMetadata(bool forceRefresh, bool includeBlocked = false)
+        {
+            var currentlyLoaded = GetLoadedMetadata(includeBlocked);
+            if (!forceRefresh && !_servicesReset && !currentlyLoaded.IsDefaultOrEmpty)
+            {
+                return currentlyLoaded;
+            }
+
+            LoadFunctionMetadata(forceRefresh);
+            _servicesReset = false;
+
+            return GetLoadedMetadata(includeBlocked);
+        }
+
+        internal ImmutableArray<FunctionMetadata> GetLoadedMetadata(bool includeBlocked)
+        {
+            return includeBlocked ? _functionMetadataAllArray : _functionMetadataFilteredArray;
+        }
+
+        private void ResetScriptHostServices(object sender, EventArgs e)
+        {
+            _functionProviders = _serviceProvider?.GetService<IEnumerable<IFunctionProvider>>();
+            _isHttpWorker = _serviceProvider?.GetService<IOptions<HttpWorkerOptions>>()?.Value?.Description != null;
+            _scriptOptions = _serviceProvider?.GetService<IOptions<ScriptJobHostOptions>>();
+
+            // Resetting the logger switches the logger scope to Script Host level,
+            // also making the logs available to Application Insights
+            _logger = _serviceProvider?.GetService<ILoggerFactory>().CreateLogger(LogCategories.Startup);
+            _servicesReset = true;
+        }
+
         /// <summary>
         /// Read all functions and populate function metadata.
         /// </summary>
-        internal ImmutableArray<FunctionMetadata> LoadFunctionMetadata()
+        internal void LoadFunctionMetadata(bool forceRefresh = false)
         {
-            ICollection<string> functionsWhiteList = _scriptOptions.Value.Functions;
+            ICollection<string> functionsWhiteList = _scriptOptions?.Value?.Functions;
             _logger.FunctionMetadataManagerLoadingFunctionsMetadata();
 
-            List<FunctionMetadata> functionMetadataList = _functionMetadataProvider.GetFunctionMetadata().ToList();
-            _functionErrors = _functionMetadataProvider.FunctionErrors.ToDictionary(kvp => kvp.Key, kvp => (ICollection<string>)kvp.Value.ToList());
+            var immutableFunctionMetadata = _functionMetadataProvider.GetFunctionMetadata(forceRefresh);
+            var functionMetadataList = new List<FunctionMetadata>();
+
+            if (!immutableFunctionMetadata.IsDefaultOrEmpty)
+            {
+                functionMetadataList.AddRange(immutableFunctionMetadata);
+            }
+
+            if (!_functionMetadataProvider.FunctionErrors?.IsEmpty ?? false)
+            {
+                _functionErrors = _functionMetadataProvider.FunctionErrors.ToDictionary(kvp => kvp.Key, kvp => (ICollection<string>)kvp.Value.ToList());
+            }
+
+            // Add metadata and errors from any additional function providers
+            if (_functionProviders != null && _functionProviders.Any())
```

Comment:
nit: `_functionProviders?.Any()`

Reply:
It'd have to be `_functionProviders?.Any() ?? false`. Not sure which is better. :)

Reply:
I'd just remove this if block entirely, and compress your two Add methods into a single LoadCustomProviderFunctions method that does the null check and adds metadata/errors